File: demo/java-demo/src/main/java/com/alibaba/testable/demo/service/DemoMatcherService.java
package com.alibaba.testable.demo.service;


import com.alibaba.testable.demo.model.BlackBox;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

/**
 * @author flin
 */
@Service
public class DemoMatcherService {

    /**
     * Method to be mocked
     */
    private void methodToBeMocked() {
        // pretend to have some code here
    }

    /**
     * Method to be mocked
     */
    private void methodToBeMocked(Object a1, Object a2) {
        // pretend to have some code here
    }

    /**
     * Method to be mocked
     */
    private void methodToBeMocked(Object[] a) {
        // pretend to have some code here
    }

    public void callMethodWithoutArgument() {
        methodToBeMocked();
    }

    public void callMethodWithNumberArguments() {
        // named variable and lambda variable will be recorded as different type
        // should have them both in test case
        List<Float> floatList = new ArrayList<>();
        floatList.add(1.0F);
        floatList.add(2.0F);
        methodToBeMocked(1, 2);
        methodToBeMocked(1L, 2.0);
        Long[] longArray = new Long[]{1L, 2L};
        methodToBeMocked(new ArrayList<Integer>(){{ add(1); }}, new HashSet<Float>(){{ add(1.0F); }});
        methodToBeMocked(1.0, new HashMap<Integer, Float>(2){{ put(1, 1.0F); }});
        methodToBeMocked(floatList, floatList);
        methodToBeMocked(longArray);
        methodToBeMocked(new Double[]{1.0, 2.0});
    }

    public void callMethodWithStringArgument() {
        methodToBeMocked("hello", "world");
        methodToBeMocked("testable", "mock");
        methodToBeMocked(new String[]{"demo"});
    }

    public void callMethodWithObjectArgument() {
        methodToBeMocked(new BlackBox("hello"), new BlackBox("world"));
        methodToBeMocked(new BlackBox("demo"), null);
        methodToBeMocked(null, new BlackBox("demo"));
    }

}

File: demo/kotlin-demo/src/main/kotlin/com/alibaba/testable/demo/service/DemoMatcherService.kt
package com.alibaba.testable.demo.service

import com.alibaba.testable.demo.model.BlackBox
import org.springframework.stereotype.Service
import java.util.*

/**
 * @author flin
 */
@Service
class DemoMatcherService {
    /**
     * Method to be mocked
     */
    private fun methodToBeMocked() {
        // pretend to have some code here
    }

    /**
     * Method to be mocked
     */
    private fun methodToBeMocked(a1: Any?, a2: Any?) {
        // pretend to have some code here
    }

    /**
     * Method to be mocked
     */
    private fun methodToBeMocked(a: Array<out Any>) {
        // pretend to have some code here
    }

    fun callMethodWithoutArgument() {
        methodToBeMocked()
    }

    fun callMethodWithNumberArguments() {
        // named variable and lambda variable will be recorded as different type
        // should have them both in test case
        val floatList: MutableList<Float> = ArrayList()
        floatList.add(1.0f)
        floatList.add(2.0f)
        val longArray = arrayOf(1L, 2L)
        methodToBeMocked(1, 2)
        methodToBeMocked(1L, 2.0)
        methodToBeMocked(listOf(1), setOf(1.0f))
        methodToBeMocked(1.0, mapOf(1 to 1.0f))
        methodToBeMocked(floatList, floatList)
        methodToBeMocked(longArray)
        methodToBeMocked(arrayOf(1.0, 2.0))
    }

    fun callMethodWithStringArgument() {
        methodToBeMocked("hello", "world")
        methodToBeMocked("testable", "mock")
        methodToBeMocked(arrayOf("demo"))
    }

    fun callMethodWithObjectArgument() {
        methodToBeMocked(BlackBox("hello"), BlackBox("world"))
        methodToBeMocked(BlackBox("demo"), null)
        methodToBeMocked(null, BlackBox("demo"))
    }
}
